File: 309-best-time-to-buy-and-sell-stock-with-cooldown/best-time-to-buy-and-sell-stock-with-cooldown.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        mapper = {}
        
        def dfs(i, task):
            if i >= len(prices): return 0
            if (i, task) in mapper: return mapper[(i, task)]

            if task:
                buying = dfs(i+1, not task) - prices[i]
                cooldown = dfs(i+1, task)
                mapper[(i, task)] = max(buying, cooldown)
            else:
                selling = dfs(i+2, not task) + prices[i]
                cooldown = dfs(i+1, task)
                mapper[(i, task)] = max(selling, cooldown)

            return mapper[(i, task)]
        
        return dfs(0, True)

File: 309-best-time-to-buy-and-sell-stock-with-cooldown/best-time-to-buy-and-sell-stock-with-cooldown.cpp
class Solution {
public:
    unordered_map<string, int> mapper;  

    int dfs(int i, bool task, vector<int>& prices){
        if (i >= prices.size()) return 0;
        string key = to_string(i) + "_" + (task ? "t":"f");
        if (mapper.contains(key)) return mapper[key];
        if (task){
            int buying = dfs(i+1, !task, prices) - prices[i];
            int cooldown = dfs(i+1, task, prices);
            mapper[key] = max(buying, cooldown);
        }
        else{
            int selling = dfs(i+2, !task, prices) + prices[i];
            int cooldown = dfs(i+1, task, prices);
            mapper[key] = max(selling, cooldown);
        }
        return mapper[key];
    }

    int maxProfit(vector<int>& prices) {
        return dfs(0, true, prices);
    }
};